File: go/supermarket/receipt.go
package supermarket

import "sort"

type Receipt struct {
    items []ReceiptItem
    discounts []Discount
}

type ReceiptItem struct {
    product Product
    quantity float64
    price float64
    totalPrice float64
}

func NewReceipt() *Receipt {
    var r Receipt
    r.items = []ReceiptItem{}
    r.discounts = []Discount{}
    return &r
}

func (r *Receipt) totalPrice() float64 {
    var total float64 = 0
    for _, item := range r.items {
        total += item.totalPrice
    }
    for _, discount := range r.discounts {
        total += discount.discountAmount
    }
    return total
}


func (r *Receipt) addProduct(product Product, quantity float64, unitPrice float64, price float64) {
    r.items = append(r.items, ReceiptItem{
        product:    product,
        quantity:   quantity,
        price:      unitPrice,
        totalPrice: price,
    })
}

func (r *Receipt) addDiscount(discount Discount) {
    r.discounts = append(r.discounts, discount)
}

func (r Receipt) sortedItems() []ReceiptItem {
    sort.Slice(r.items, func(i, j int) bool {
        return r.items[i].product.name < r.items[j].product.name
    })
    return r.items
}

func (r Receipt) sortedDiscounts() []Discount {
    sort.Slice(r.discounts, func(i, j int) bool {
        return r.discounts[i].product.name < r.discounts[j].product.name
    })
    return r.discounts
}

File: python/supermarket/receipt.py
from dataclasses import dataclass
from typing import List

from supermarket.product import Product
from supermarket.offer import Discount


@dataclass(frozen=True)
class ReceiptItem:
    product: Product
    quantity: float
    price: float
    total_price: float


class Receipt:
    def __init__(self) -> None:
        self._items: List[ReceiptItem] = []
        self._discounts: List[Discount] = []

    def total_price(self) -> float:
        return sum(i.total_price for i in self.items) + sum(
            d.discount_amount for d in self.discounts
        )

    def add_product(
        self, product: Product, quantity: float, price: float, total_price: float
    ) -> None:
        # TODO: pass the object instead of so many arguments
        self._items.append(ReceiptItem(product, quantity, price, total_price))

    def add_discount(self, discount: Discount) -> None:
        self._discounts.append(discount)

    @property
    def items(self) -> List[ReceiptItem]:
        return self._items[:]

    @property
    def discounts(self) -> List[Discount]:
        return self._discounts[:]
